[Arm64] Add basic ARMv8.1 Atomics

sdmaclea opened the pull request:
Add cas*, ldadd*, and swp*

@briansull @dotnet/arm64-contrib These are the 8.1 atomics that will be useful with current Interlocked operations with and without explicit memory ordering

sdmaclea commented:
So far I have compiled and checked most of the encodings with our internal arm64 disassembler.  

I planned to visually verify again and run all the encoding through our disassembler again just to make sure.

> Have you ever used CodeGen::genArm64EmitterUnitTests?

I have not, but I am aware of it, and added unit tests for it.

I plan to add the codegenerators soon.  I will add the new units tests too.  Perhaps I will even try genArm64EmitterUnitTests, but I do not know if it will work with my build infrastructure.

briansull commented:
These are new instructions not documented in the **ArmV8.pdf** that we have been using up to now.
I have located and downloaded the new **ARM 8.1 supplement** that describes the instructions.

It can be downloaded from here:

https://developer.arm.com/products/architecture/a-profile/docs/ddi0557/a/arm-architecture-reference-manual-supplement-armv81-for-armv8-a-architecture-profile

sdmaclea commented:
They were added to the ARMv8 ARM 3/2017 
`31 March 2017` `B.a` `Non-Confidential v8.1 EAC, v8.2 Beta` `Initial release incorporating ARMv8.1 and ARMv8.2.`

briansull commented:
If I download a new  **ArmV8.pdf**  will it have these instructions, or are these just in the 8.1 supplement

sdmaclea commented:
I don't know what is on the ARM website.  The latest EAC ARMv8 ARM Version C.a 12/2017 has 8.1, 8.2, and 8.3 extensions included.

briansull commented:
Is there a reason that you aren't adding the ST<OP> forms?  i.e stadd*

briansull commented:
It also seems like the ldadd and stadd instruction could be used to implement i++ and ++i using a single instruction (instead of three instructions),  would that be a bad idea or would it be acceptable.
Often times we want to generate smaller code rather than larger code, even it it runs a bit slower.

sdmaclea commented:
> Is there a reason that you aren't adding the ST forms? i.e stadd*

stadd is the preferred disassembly for ldadd with target WZR.
```
STADD <Ws>, [<Xn|SP>]
is equivalent to
LDADD <Ws>, WZR, [<Xn|SP>]
```

I didn't add because:
+ I wasn't sure it was necessary, 
+ I wasn't sure what complications it would cause.
+ It wasn't immediately obvious to me how to accomplish it

Please advise.

briansull commented:
You can simply add this code to emitIns_R_R(...)

```
        case INS_staddb:
        case INS_staddab:
        case INS_staddalb:
        case INS_staddlb:
        case INS_staddh:
        case INS_staddah:
        case INS_staddalh:
        case INS_staddlh:
        case INS_stadd:
        case INS_stadda:
        case INS_staddal:
        case INS_staddl:
              // call helper to swap the opcode from st* to ld*
              instruction ldIns = toLdForm(ins);
             return emitIns_R_R_R(ldIns, size, reg1, REG_ZR, reg2);
```

briansull commented on src/jit/instrsarm64.h:
```diff
@@ -708,6 +708,114 @@ INST1(sturb,   "sturb",  0,ST, IF_LS_2C,  0x38000000)
+INST1(ldaddb,  "ldaddb",  0, LD|ST, IF_LS_3E,  0x38200000)
+                                   //  ldaddb   Wm, Wt, [Xn]        LS_3E  00111000001mmmmm 000000nnnnnttttt   3820 0000   Rm Rt Rn ARMv8.1 LSE Atomics
+
+INST1(ldaddab, "ldaddab", 0, LD|ST, IF_LS_3E,  0x38A00000)
```
You could add the store forms for these instruction before or afteer each of the load forms so that the implementations of:
`toLdForm(ins) `and `toStForm(ins) `would be trivial.

sdmaclea commented:
> It also seems like the ldadd and stadd instruction could be used to implement i++ and ++i using a single instruction (instead of three instructions), would that be a bad idea or would it be acceptable.
Often times we want to generate smaller code rather than larger code, even it it runs a bit slower.

It may not generate smaller code.  Given the limited addressing forms, you would need to calculate the address.  Also you need to have a `1` in a register.  So it would likely be three instructions.

If `i` was promoted to a register, it would only take one instruction.

From a performance perspective, if you are going to memory anyway it should be efficient (w/o memory ordering).

sdmaclea commented:
Added `stadd*` forms.  Added emiterr code and tests.

I tried to run genArm64EmitterUnitTests on linux-arm64, but it didn't work.  I suspect I need to run it in x64.  I'll have to rework my workflow...

sdmaclea commented:
LATE_DISAM is off by default.  Trying to enable it shows that it doesn't compile on Linux.
```
/home/sdmaclea/git/coreclr/src/jit/disasm.h:41:10: fatal error: 'msvcdis.h' file not found
#include "msvcdis.h"
```

So I can't run `genArm64EmitterUnitTests`

BruceForstall commented:
genArm64EmitterUnitTests basically generates a bunch of instructions and then calls the disassembler to output them, at which point you need to see, manually, that what was output is what you expect. You could do this on Windows x64 using the arm64 altjit and the msvcdis disassembler (although, to be honest, I don't remember if all the appropriate header files are public). For Linux, if we ever build an arm64 coredistools disassembler package, we could build it to use that. Barring that, you might be able to turn off the disassembly dump, and still use genArm64EmitterUnitTests to inject the instructions. Run under your favorite debugger with COMPlus_JitHalt, then use your debugger to disassemble the instructions.

sdmaclea commented:
I was able to get `genArm64EmitterUnitTests()` to be useful on linux.  

I hacked `genArm64EmitterUnitTests()` to always run.  

Ran crosscomponent crossgen on a simple test.  In this case `JIT/CodeGenBringUpTests/Add1/Add1.exe`.  

Ran objdump -D -b binary -m aarch64 Add.ni.exe.  Manually looked for relevant part...
```
     91c:       08a87d49        casb    w8, w9, [x10]
     920:       08e87d49        casab   w8, w9, [x10]
     924:       08e8fd49        casalb  w8, w9, [x10]
     928:       08a8fd49        caslb   w8, w9, [x10]
     92c:       48a87d49        cash    w8, w9, [x10]
     930:       48e87d49        casah   w8, w9, [x10]
     934:       48e8fd49        casalh  w8, w9, [x10]
     938:       48a8fd49        caslh   w8, w9, [x10]
     93c:       88a87d49        cas     w8, w9, [x10]
     940:       88e87d49        casa    w8, w9, [x10]
     944:       88e8fd49        casal   w8, w9, [x10]
     948:       88a8fd49        casl    w8, w9, [x10]
     94c:       c8a87d49        cas     x8, x9, [x10]
     950:       c8e87d49        casa    x8, x9, [x10]
     954:       c8e8fd49        casal   x8, x9, [x10]
     958:       c8a8fd49        casl    x8, x9, [x10]
     95c:       38280149        ldaddb  w8, w9, [x10]
     960:       38a80149        ldaddab w8, w9, [x10]
     964:       38e80149        ldaddalb        w8, w9, [x10]
     968:       38680149        ldaddlb w8, w9, [x10]
     96c:       78280149        ldaddh  w8, w9, [x10]
     970:       78a80149        ldaddah w8, w9, [x10]
     974:       78e80149        ldaddalh        w8, w9, [x10]
     978:       78680149        ldaddlh w8, w9, [x10]
     97c:       b8280149        ldadd   w8, w9, [x10]
     980:       b8a80149        ldadda  w8, w9, [x10]
     984:       b8e80149        ldaddal w8, w9, [x10]
     988:       b8680149        ldaddl  w8, w9, [x10]
     98c:       f8280149        ldadd   x8, x9, [x10]
     990:       f8a80149        ldadda  x8, x9, [x10]
     994:       f8e80149        ldaddal x8, x9, [x10]
     998:       f8680149        ldaddl  x8, x9, [x10]
     99c:       38288149        swpb    w8, w9, [x10]
     9a0:       38a88149        swpab   w8, w9, [x10]
     9a4:       38e88149        swpalb  w8, w9, [x10]
     9a8:       38688149        swplb   w8, w9, [x10]
     9ac:       78288149        swph    w8, w9, [x10]
     9b0:       78a88149        swpah   w8, w9, [x10]
     9b4:       78e88149        swpalh  w8, w9, [x10]
     9b8:       78688149        swplh   w8, w9, [x10]
     9bc:       b8288149        swp     w8, w9, [x10]
     9c0:       b8a88149        swpa    w8, w9, [x10]
     9c4:       b8e88149        swpal   w8, w9, [x10]
     9c8:       b8688149        swpl    w8, w9, [x10]
     9cc:       f8288149        swp     x8, x9, [x10]
     9d0:       f8a88149        swpa    x8, x9, [x10]
     9d4:       f8e88149        swpal   x8, x9, [x10]
     9d8:       f8688149        swpl    x8, x9, [x10]
     9dc:       3828015f        staddb  w8, [x10]
     9e0:       3868015f        staddlb w8, [x10]
     9e4:       7828015f        staddh  w8, [x10]
     9e8:       7868015f        staddlh w8, [x10]
     9ec:       b828015f        stadd   w8, [x10]
     9f0:       b868015f        staddl  w8, [x10]
     9f4:       f828015f        stadd   x8, [x10]
     9f8:       f868015f        staddl  x8, [x10]
```

sdmaclea commented:
The unit test looked good.  I think this can be merged.

BruceForstall commented:
Thanks for adding the unit tests.

How should we think about versioning for arm64? Presumably these 8.1 instructions won't run on older chips that we want to support. Should the instructions themselves be `#ifdef`ed? Should we demarcate the instructions similar to `INS_FIRST_AVX_INSTRUCTION` and `INS_LAST_AVX_INSTRUCTION` on xarch, and then add a `IsArmV81Instruction()` function? I see there already is `InstructionSet_Simd_v81`; should this be used to gate usage of these instructions?

sdmaclea commented:
These correspond to `JIT_FLAG_HAS_ARM64_ATOMICS` & `CORJIT_FLAG_HAS_ARM64_ATOMICS`

Eventually we may want an assert somewhere. I was thinking about putting it in the emit code, but hadn't decided yet.  I was wondering if we might allow the instructions be generated in certain circumstances.  (For instance during crossgen if the generated code is runtime gated)

BruceForstall commented:
ok. I'm mostly worried about generation of new instructions "leaking" into code paths intended for older processors. I'll go ahead and merge this, since you're already thinking about this. But before we actually generate these, we should make sure to have appropriate checks/asserts in place.
